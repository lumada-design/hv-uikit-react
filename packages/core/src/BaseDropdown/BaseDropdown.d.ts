import * as React from "react";
import { StandardProps } from "@material-ui/core";
import { PopperProps } from "@material-ui/core/Popper";

export interface HvBaseDropdownProps
  extends StandardProps<React.HTMLAttributes<HTMLDivElement>, HvBaseDropdownClassKey> {
  /**
   * Header placeholder.
   */
  placeholder?: string | React.ReactNode;
  /**
   * If ´true´ the dropdown is disabled unable to be interacted, if ´false´ it is enabled.
   */
  disabled?: boolean;
  /**
   * Disable the portal behavior.
   * The children stay within it's parent DOM hierarchy.
   */
  disablePortal?: boolean;
  /**
   * If ´true´ the dropdown starts opened if ´false´ it starts closed.
   */
  expanded?: boolean;
  /**
   * An object containing props to be wired to the popper component.
   */
  popperProps?: Partial<PopperProps>;
  /**
   * Node to be rendered.
   */
  children?: React.ReactNode;
  /**
   * Placement of the dropdown.
   */
  placement?: "left" | "right";
  /**
   * Replacement for the header component.
   */
  component?: React.ReactNode;
  /**
   * Adornment to replace the default arrows.
   */
  adornment?: React.ReactNode;
  /**
   * When dropdown changes the expanded state.
   */
  onToggle?: (event: Event) => void;
  /**
   * When user click outside the open container.
   */
  onClickOutside?: (event: Event) => void;
  /**
   * When expanded dropdown changes position.
   */
  onFlip?: (flipped: boolean) => void;
}

export type HvBaseDropdownClassKey =
  | "root"
  | "header"
  | "headerOpen"
  | "headerOpenUp"
  | "headerOpenDown"
  | "headerDisabled"
  | "arrow"
  | "selection"
  | "truncate"
  | "selectionDisabled"
  | "panel"
  | "inputExtensionOpen"
  | "inputExtensionLeftPosition"
  | "inputExtensionOpenShadow"
  | "inputExtensionFloatRight"
  | "inputExtensionFloatLeft";

export default function HvBaseDropdown(props: HvBaseDropdownProps): JSX.Element | null;
