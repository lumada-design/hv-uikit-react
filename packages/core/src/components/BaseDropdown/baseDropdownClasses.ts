import { getClasses } from "@core/utils";

export interface HvBaseDropdownClasses {
  root?: string;
  rootDisabled?: string;
  anchor?: string;
  container?: string;
  header?: string;
  headerOpen?: string;
  headerOpenUp?: string;
  headerOpenDown?: string;
  headerDisabled?: string;
  headerReadOnly?: string;
  arrow?: string;
  selection?: string;
  placeholder?: string;
  selectionDisabled?: string;
  panel?: string;
  inputExtensionOpen?: string;
  inputExtensionLeftPosition?: string;
  inputExtensionOpenShadow?: string;
  inputExtensionFloatRight?: string;
  inputExtensionFloatLeft?: string;
}

const classKeys: (keyof HvBaseDropdownClasses)[] = [
  "root",
  "rootDisabled",
  "anchor",
  "container",
  "header",
  "headerOpen",
  "headerOpenUp",
  "headerOpenDown",
  "headerDisabled",
  "headerReadOnly",
  "arrow",
  "selection",
  "placeholder",
  "selectionDisabled",
  "panel",
  "inputExtensionOpen",
  "inputExtensionLeftPosition",
  "inputExtensionOpenShadow",
  "inputExtensionFloatRight",
  "inputExtensionFloatLeft",
];

const baseDropdownClasses = getClasses(classKeys, "HvBaseDropdown");

export default baseDropdownClasses;
