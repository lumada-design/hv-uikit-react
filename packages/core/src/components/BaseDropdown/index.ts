export { default as baseDropdownClasses } from "./baseDropdownClasses";
export * from "./baseDropdownClasses";
export * from "./BaseDropdown";
