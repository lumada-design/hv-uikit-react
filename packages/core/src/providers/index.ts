export * from "./Provider";
export * from "./ThemeProvider";
