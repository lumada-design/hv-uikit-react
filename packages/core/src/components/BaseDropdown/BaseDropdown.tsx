import React, { useMemo, useState, useCallback } from "react";
import { createPortal } from "react-dom";
import { clsx } from "clsx";
import BaseDropdownContext from "./BaseDropdownContext";
import {
  ClickAwayListener,
  PopperPlacementType,
  PopperProps,
} from "@mui/material";
import { useControlled, useForkRef, useTheme, useUniqueId } from "@core/hooks";
import { isKeypress, keyboardCodes, setId } from "@core/utils";
import { getFirstAndLastFocus } from "@core/utils/focusableElementFinder";
import { HvBaseProps } from "@core/types";
import {
  StyledAnchor,
  StyledContainer,
  StyledDropDownXS,
  StyledDropUpXS,
  StyledExtension,
  StyledHeaderRoot,
  StyledPanel,
  StyledPlaceholder,
  StyledRoot,
  StyledSelection,
} from "./BaseDropdown.styles";
import { usePopper } from "react-popper";
import { detectOverflow, ModifierArguments, Options } from "@popperjs/core";
import baseDropdownClasses, {
  HvBaseDropdownClasses,
} from "./baseDropdownClasses";

const { Tab, Enter, Esc, Space, ArrowDown } = keyboardCodes;

export interface HvBaseDropdownProps
  extends HvBaseProps<HTMLDivElement, "placeholder"> {
  /**
   * The role of the element that triggers the popup.
   *
   * Defaults to "combobox" if `component` and the default
   * "textbox" header is used, undefined otherwise.
   */
  role?: string;
  /**
   * Header placeholder.
   */
  placeholder?: string | React.ReactNode;
  /**
   * If `true` the dropdown is disabled unable to be interacted, if `false` it is enabled.
   */
  disabled?: boolean;
  /**
   * If `true` the dropdown will be in read only mode, unable to be interacted.
   */
  readOnly?: boolean;
  /**
   * Disable the portal behavior.
   * The children stay within it's parent DOM hierarchy.
   */
  disablePortal?: boolean;
  /**
   * If `true` the dropdown width depends size of content if `false` the width depends on the header size.
   * Defaults to `false`.
   */
  variableWidth?: boolean;
  /**
   * If `true` the dropdown starts opened if `false` it starts closed.
   */
  expanded?: boolean;
  /**
   * When uncontrolled, defines the initial expanded state.
   */
  defaultExpanded?: boolean;
  /**
   * An object containing props to be wired to the popper component.
   */
  popperProps?: Partial<PopperProps>;
  /**
   * Placement of the dropdown.
   */
  placement?: "left" | "right";
  /**
   * Replacement for the header component.
   */
  component?: React.ReactNode;
  /**
   * Adornment to replace the default arrows.
   */
  adornment?: React.ReactNode;
  /**
   * When dropdown changes the expanded state.
   */
  onToggle?: (event: Event, open: boolean) => void;
  /**
   * When user click outside the open container.
   */
  onClickOutside?: (event: Event) => void;
  /**
   * Callback called when the dropdown is opened and ready,
   * commonly used to set focus to the content.
   */
  onContainerCreation?: (container: HTMLElement | null) => void;
  /**
   * Attributes applied to the dropdown header element.
   */
  dropdownHeaderProps?: React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  >;
  /**
   * Pass a ref to the dropdown header element.
   */
  dropdownHeaderRef?: React.Ref<any>;
  /**
   * A Jss Object used to override or extend the component styles applied.
   */
  classes?: HvBaseDropdownClasses;
}

export const HvBaseDropdown = ({
  id,
  className,
  classes,
  children,
  role,
  placeholder,
  component,
  adornment,
  expanded,
  dropdownHeaderProps,
  defaultExpanded = false,
  disabled = false,
  readOnly = false,
  disablePortal = false,
  variableWidth = false,
  placement = "right",
  popperProps = {},
  dropdownHeaderRef: dropdownHeaderRefProp,
  onToggle,
  onClickOutside,
  onContainerCreation,
  ...others
}: HvBaseDropdownProps) => {
  const { rootId } = useTheme();

  const [isOpen, setIsOpen] = useControlled(expanded, Boolean(defaultExpanded));

  const [referenceElement, setReferenceElement] = useState<HTMLElement | null>(
    null
  );
  const [popperElement, setPopperElement] = useState<HTMLElement | null>(null);
  const [popperMaxSize, setPopperMaxSize] = useState<{
    width?: number;
    height?: number;
  }>({});

  const handleDropdownHeaderRefProp = useForkRef(
    dropdownHeaderRefProp,
    dropdownHeaderProps?.ref
  );
  const handleDropdownHeaderRef = useForkRef(
    setReferenceElement,
    handleDropdownHeaderRefProp
  );

  const ariaRole = role || (component == null ? "combobox" : undefined);

  const elementId = useUniqueId(id, "hvbasedropdown");

  const bottom: PopperPlacementType =
    placement && `bottom-${placement === "right" ? "start" : "end"}`;

  const extensionWidth = referenceElement
    ? referenceElement?.offsetWidth
    : "inherit";

  const { modifiers: popperPropsModifiers = [], ...otherPopperProps } =
    popperProps;

  const onFirstUpdate = useCallback(() => {
    if (onContainerCreation) onContainerCreation(popperElement);
  }, [onContainerCreation, popperElement]);

  const widthCalculator = useCallback(
    ({ state }: ModifierArguments<Options>) => {
      state.styles.popper.width = `${state.rects.reference.width}px`;
    },
    []
  );

  const widthCalculatorEffect = useCallback(
    ({ state }: ModifierArguments<Options>) => {
      state.elements.popper.style.width = `${
        (state.elements.reference as any).offsetWidth
      }px`;
    },
    []
  );

  const applyMaxSizeCalculator = useCallback(
    ({ state }: ModifierArguments<Options>) => {
      // The `maxSize` modifier provides this data
      const { width, height } = state.modifiersData.maxSize;
      if (width !== popperMaxSize?.width || height !== popperMaxSize?.height) {
        setPopperMaxSize({ width, height });
      }

      state.styles.popper = {
        ...state.styles.popper,
        maxWidth: `${width}px`,
        maxHeight: `${height}px`,
      };
    },
    [popperMaxSize]
  );

  const maxSizeCalculator = useCallback(
    ({ state, name, options }: ModifierArguments<Options>) => {
      const overflow = detectOverflow(state, options);

      const x = state.modifiersData.preventOverflow?.x || 0;
      const y = state.modifiersData.preventOverflow?.y || 0;

      const popperWidth = state.rects.popper.width;
      const popperHeight = state.rects.popper.height;

      const basePlacement = state.placement.split("-")[0];

      const widthProp = basePlacement === "left" ? "left" : "right";
      const heightProp = basePlacement === "top" ? "top" : "bottom";

      state.modifiersData[name] = {
        width: popperWidth - overflow[widthProp] - x,
        height: popperHeight - overflow[heightProp] - y,
      };
    },
    []
  );

  const modifiers: any = useMemo(
    () => [
      {
        name: "variableWidth",
        enabled: !variableWidth,
        phase: "beforeWrite",
        requires: ["computeStyles"],
        fn: widthCalculator,
        effect: widthCalculatorEffect,
      },
      {
        name: "maxSize",
        enabled: true,
        phase: "main",
        requiresIfExists: ["offset", "preventOverflow", "flip"],
        fn: maxSizeCalculator,
      },
      {
        name: "applyMaxSize",
        enabled: true,
        phase: "beforeWrite",
        requires: ["maxSize"],
        fn: applyMaxSizeCalculator,
      },
      ...popperPropsModifiers,
    ],
    [
      maxSizeCalculator,
      applyMaxSizeCalculator,
      popperPropsModifiers,
      variableWidth,
      widthCalculator,
      widthCalculatorEffect,
    ]
  );

  const { styles: popperStyles, attributes } = usePopper(
    referenceElement,
    popperElement,
    {
      placement: bottom,
      modifiers,
      onFirstUpdate,
      ...otherPopperProps,
    }
  );

  let popperPlacement = "bottom";
  if (attributes.popper) {
    popperPlacement = attributes.popper["data-popper-placement"];
  }

  const handleToggle = useCallback(
    (event) => {
      if (event && !isKeypress(event, Tab)) {
        event.preventDefault();
      }

      // We are checking specifically for false because if "isKeypress" returns true or undefined it should continue
      const notControlKey = [Tab, Enter, Esc, ArrowDown, Space].every(
        (key) => isKeypress(event, key) === false
      );

      const ignoredCombinations =
        (isKeypress(event, Esc) && !isOpen) ||
        (isKeypress(event, ArrowDown) && isOpen) ||
        (isKeypress(event, Tab) && !isOpen);

      if (disabled || notControlKey || ignoredCombinations) return;

      const newOpen = !isOpen;

      /* If about to close focus on the header component. */
      const focusHeader = () => {
        if (!newOpen) {
          // Focus-ring won't be visible even if using the keyboard:
          // https://github.com/WICG/focus-visible/issues/88
          referenceElement?.focus({ preventScroll: true });
        }

        return newOpen;
      };
      setIsOpen(focusHeader());

      onToggle?.(event, newOpen);
    },
    [isOpen, disabled, setIsOpen, onToggle, referenceElement]
  );

  const headerComponent = (() => {
    if (component) {
      return React.cloneElement(component as React.ReactElement, {
        ref: handleDropdownHeaderRef,
      });
    }

    return (
      <StyledHeaderRoot
        id={setId(id, "header")}
        className={clsx(
          baseDropdownClasses.header,
          classes?.header,
          disabled &&
            clsx(baseDropdownClasses.headerDisabled, classes?.headerDisabled),
          readOnly &&
            clsx(baseDropdownClasses.headerReadOnly, classes?.headerReadOnly),
          isOpen && clsx(baseDropdownClasses.headerOpen, classes?.headerOpen),
          isOpen &&
            popperPlacement.includes("top") &&
            clsx(baseDropdownClasses.headerOpenUp, classes?.headerOpenUp),
          isOpen &&
            popperPlacement.includes("bottom") &&
            clsx(baseDropdownClasses.headerOpenDown, classes?.headerOpenDown)
        )}
        $disabled={disabled}
        $readOnly={readOnly}
        $opened={isOpen}
        $openedUp={isOpen && popperPlacement.includes("top")}
        $openedDown={isOpen && popperPlacement.includes("bottom")}
        role={ariaRole === "combobox" ? "textbox" : undefined}
        style={disabled || readOnly ? { pointerEvents: "none" } : undefined}
        aria-controls={
          isOpen ? setId(elementId, "children-container") : undefined
        }
        aria-label={others["aria-label"] ?? undefined}
        aria-labelledby={others["aria-labelledby"] ?? undefined}
        // Removes the element from the navigation sequence for keyboard focus if disabled
        tabIndex={disabled ? -1 : 0}
        ref={handleDropdownHeaderRef}
        {...dropdownHeaderProps}
      >
        <StyledSelection
          className={clsx(baseDropdownClasses.selection, classes?.selection)}
        >
          {placeholder && typeof placeholder === "string" ? (
            <StyledPlaceholder
              className={clsx(
                baseDropdownClasses.placeholder,
                classes?.placeholder,
                disabled &&
                  clsx(
                    baseDropdownClasses.selectionDisabled,
                    classes?.selectionDisabled
                  )
              )}
              $disabled={disabled}
              variant="body"
            >
              {placeholder}
            </StyledPlaceholder>
          ) : (
            placeholder
          )}
        </StyledSelection>
        {adornment ||
          (isOpen ? (
            <StyledDropUpXS
              iconSize="XS"
              className={clsx(baseDropdownClasses.arrow, classes?.arrow)}
            />
          ) : (
            <StyledDropDownXS
              iconSize="XS"
              className={clsx(baseDropdownClasses.arrow, classes?.arrow)}
              $disabled={disabled}
            />
          ))}
      </StyledHeaderRoot>
    );
  })();

  const containerComponent = (() => {
    /**
     *  Handle keyboard inside children container.
     */
    const handleContainerKeyDown = (event) => {
      if (isKeypress(event, Esc)) {
        handleToggle(event);
      }
      if (isKeypress(event, Tab) && !event.shiftKey) {
        const focusList = getFirstAndLastFocus(popperElement);
        if (document.activeElement === focusList?.last) {
          event.preventDefault();
          focusList?.first?.focus();
        }
      }
    };

    const handleOutside = (event) => {
      const isButtonClick = referenceElement?.contains(event.target);
      if (!isButtonClick) {
        onClickOutside?.(event);
        setIsOpen(false);
        onToggle?.(event, false);
      }
    };

    const container = (
      <StyledContainer
        role="tooltip"
        ref={setPopperElement}
        className={clsx(baseDropdownClasses.container, classes?.container)}
        style={popperStyles.popper}
        {...attributes.popper}
      >
        <ClickAwayListener onClickAway={handleOutside}>
          <div onKeyDown={handleContainerKeyDown}>
            {popperPlacement.includes("bottom") && (
              <StyledExtension
                style={{ width: extensionWidth }}
                className={clsx(
                  baseDropdownClasses.inputExtensionOpen,
                  classes?.inputExtensionOpen,
                  popperPlacement.includes("end") &&
                    clsx(
                      baseDropdownClasses.inputExtensionLeftPosition,
                      classes?.inputExtensionLeftPosition
                    )
                )}
                $leftPosition={popperPlacement.includes("end")}
                $openShadow={false}
                $floatLeft={false}
                $floatRight={false}
              />
            )}
            <BaseDropdownContext.Provider value={popperMaxSize}>
              <StyledPanel
                id={setId(elementId, "children-container")}
                className={clsx(baseDropdownClasses.panel, classes?.panel)}
                $popperPlacement={
                  popperPlacement.includes("top") ? "top" : "bottom"
                }
              >
                {children}
              </StyledPanel>
            </BaseDropdownContext.Provider>
            {popperPlacement.includes("top") && (
              <StyledExtension
                style={{ width: extensionWidth }}
                className={clsx(
                  baseDropdownClasses.inputExtensionOpen,
                  classes?.inputExtensionOpen,
                  baseDropdownClasses.inputExtensionOpenShadow,
                  classes?.inputExtensionOpenShadow,
                  popperPlacement.includes("end") &&
                    clsx(
                      baseDropdownClasses.inputExtensionFloatRight,
                      classes?.inputExtensionFloatRight
                    ),
                  popperPlacement.includes("start") &&
                    clsx(
                      baseDropdownClasses.inputExtensionFloatLeft,
                      classes?.inputExtensionFloatLeft
                    )
                )}
                $leftPosition={false}
                $openShadow={true}
                $floatLeft={popperPlacement.includes("start")}
                $floatRight={popperPlacement.includes("end")}
              />
            )}
          </div>
        </ClickAwayListener>
      </StyledContainer>
    );

    if (disablePortal) return container;

    return createPortal(
      container,
      document.getElementById(rootId || "") || document.body
    );
  })();

  return (
    <StyledRoot className={clsx(baseDropdownClasses.root, classes?.root)}>
      <StyledAnchor
        id={id}
        role={ariaRole}
        aria-expanded={!!isOpen}
        aria-owns={isOpen ? setId(elementId, "children-container") : undefined}
        className={clsx(
          className,
          baseDropdownClasses.anchor,
          classes?.anchor,
          disabled &&
            clsx(baseDropdownClasses.rootDisabled, classes?.rootDisabled)
        )}
        $disabled={disabled}
        {...(!readOnly && {
          onKeyDown: handleToggle,
          onClick: handleToggle,
        })}
        // Removes the element from the navigation sequence for keyboard focus
        tabIndex={-1}
        {...others}
      >
        {headerComponent}
      </StyledAnchor>
      {isOpen && containerComponent}
    </StyledRoot>
  );
};
