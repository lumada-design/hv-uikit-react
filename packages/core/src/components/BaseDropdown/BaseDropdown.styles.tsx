import { CSSProperties } from "@emotion/serialize";
import styled from "@emotion/styled";
import { DropDownXS, DropUpXS } from "@hitachivantara/uikit-icons";
import { theme } from "@hitachivantara/uikit-styles";
import { HvTypography } from "components";
import { outlineStyles } from "utils";
import { transientOptions } from "utils/transientOptions";

export const StyledRoot = styled("div")({
  width: "100%",
  position: "relative",
});

export const StyledAnchor = styled(
  "div",
  transientOptions
)(({ $disabled }: { $disabled: boolean }) => ({
  display: "inline-block",
  width: "100%",

  ...($disabled && {
    cursor: "not-allowed",
    "&:focus": {
      outline: "none",
    },
  }),
}));

export const StyledHeaderRoot = styled(
  "div",
  transientOptions
)(
  ({
    $disabled,
    $readOnly,
    $opened,
    $openedUp,
    $openedDown,
  }: {
    $disabled: boolean;
    $readOnly: boolean;
    $opened: boolean;
    $openedUp: boolean;
    $openedDown: boolean;
  }) => ({
    cursor: "pointer",
    userSelect: "none",
    position: "relative",
    background: theme.colors.atmo1,
    border: `1px solid ${theme.baseDropdown.borderColor}`,
    borderRadius: "2px",
    "&:hover": {
      border: `1px solid ${theme.baseDropdown.hoverBorderColor}`,
    },
    "&:focus": {
      outline: "none",
    },
    "&:focus-visible": {
      ...outlineStyles,
      border: `1px solid ${theme.baseDropdown.hoverBorderColor}`,
    },

    ...($disabled && {
      cursor: "not-allowed",
      pointerEvents: "none",
      border: `1px solid ${theme.baseDropdown.disabledBorderColor}`,
      background: theme.baseDropdown.disabledBackgroundColor,
      "&:hover": {
        border: `1px solid ${theme.baseDropdown.disabledBorderColor}`,
      },
    }),

    ...($readOnly && {
      cursor: "not-allowed",
      pointerEvents: "none",
      border: theme.baseDropdown.readOnlyBorder,
      background: theme.baseDropdown.readOnlyBackgroundColor,
      userSelect: "text",
      "&:focus-visible": {
        outline: "none",
        border: theme.baseDropdown.readOnlyBorder,
      },
    }),

    ...($opened && {
      border: `1px solid ${theme.baseDropdown.openBorderColor}`,
      boxShadow: theme.baseDropdown.shadow,
      "&:hover": {
        border: `1px solid ${theme.baseDropdown.openBorderColor}`,
        boxShadow: theme.baseDropdown.shadow,
      },
    }),

    ...($openedUp && {
      borderRadius: "0px 0px 2px 2px",
    }),

    ...($openedDown && {
      borderRadius: "2px 2px 0px 0px",
    }),
  })
);

export const StyledSelection = styled("div")({
  display: "flex",
  alignItems: "center",
  height: "30px",
  padding: "0px 30px 0px 10px",
});

export const StyledPlaceholder = styled(
  HvTypography,
  transientOptions
)(({ $disabled }: { $disabled: boolean }) => ({
  display: "block",
  overflow: "hidden",
  textOverflow: "ellipsis",
  whiteSpace: "nowrap",
  color: theme.baseDropdown.placeholderColor,
  letterSpacing: theme.baseDropdown.letterSpacing,
  lineHeight: theme.baseDropdown.lineHeight,
  fontSize: theme.baseDropdown.fontSize,
  fontWeight: theme.fontWeights.normal,

  ...($disabled && {
    color: theme.colors.atmo5,
    fontWeight: theme.fontWeights.normal,
    fontSize: theme.baseDropdown.fontSize,
    lineHeight: theme.baseDropdown.lineHeight,
    letterSpacing: theme.baseDropdown.letterSpacing,
  }),
}));

const iconsStyles: CSSProperties = {
  position: "absolute",
  pointerEvents: "none",
  top: -1,
  right: -1,
};

export const StyledDropUpXS = styled(DropUpXS)({
  ...iconsStyles,
});

export const StyledDropDownXS = styled(
  DropDownXS,
  transientOptions
)(({ $disabled }: { $disabled: boolean }) => ({
  ...iconsStyles,

  ...($disabled && {
    "& svg": {
      "& path": {
        fill: theme.colors.atmo5,
      },
    },
  }),
}));

export const StyledContainer = styled(
  "div",
  transientOptions
)(({ $zIndex }: { $zIndex: string | number }) => ({
  zIndex: $zIndex,
}));

export const StyledExtension = styled(
  "div",
  transientOptions
)(
  ({
    $leftPosition,
    $openShadow,
    $floatLeft,
    $floatRight,
    $backgroundColor,
    $shadowColor,
  }: {
    $leftPosition: boolean;
    $openShadow: boolean;
    $floatLeft: boolean;
    $floatRight: boolean;
    $backgroundColor: string;
    $shadowColor: string;
  }) => ({
    height: "10px",
    backgroundColor: $backgroundColor,

    ...($leftPosition && {
      marginLeft: "auto",
    }),

    ...($openShadow && {
      boxShadow: `0px 8px 0px ${$shadowColor}, 0px 0px 9px 0px rgba(65,65,65,.12)`,
    }),

    ...($floatLeft && {
      float: "left",
    }),

    ...($floatRight && {
      float: "right",
    }),
  })
);

export const StyledPanel = styled(
  "div",
  transientOptions
)(({ $shadowColor }: { $shadowColor: string }) => ({
  position: "relative",
  boxShadow: $shadowColor,
}));
