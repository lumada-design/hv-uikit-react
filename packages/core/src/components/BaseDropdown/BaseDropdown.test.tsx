import userEvent from "@testing-library/user-event";
import { fireEvent, render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { HvBaseDropdown } from "./BaseDropdown";
import { HvProvider } from "@core/providers";

const Main = () => (
  <HvProvider>
    <div style={{ width: 121 }}>
      <HvBaseDropdown placeholder="Placeholder..." aria-label="Main sample" />
    </div>
  </HvProvider>
);

describe("BaseDropDown", () => {
  it("should be defined", () => {
    const { container } = render(<Main />);

    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(<Main />);

    expect(container).toMatchSnapshot();
  });

  it("should open on click", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    await userEvent.click(baseDropdownHeader);

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "true");
  });

  it("should close on double click", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    await userEvent.click(baseDropdownHeader);

    // Close
    await userEvent.click(baseDropdownHeader);

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");
  });

  it("should open on Enter", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    fireEvent.keyDown(baseDropdownHeader, { key: "Enter", keyCode: 13 });

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "true");
  });

  it("should close on double Enter", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    fireEvent.keyDown(baseDropdownHeader, { key: "Enter", keyCode: 13 });

    // Close
    fireEvent.keyDown(baseDropdownHeader, { key: "Enter", keyCode: 13 });

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");
  });

  it("should open on Space", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    fireEvent.keyDown(baseDropdownHeader, { key: " ", keyCode: 32 });

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "true");
  });

  it("should close on double Space", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    fireEvent.keyDown(baseDropdownHeader, { key: " ", keyCode: 32 });

    // Close
    fireEvent.keyDown(baseDropdownHeader, { key: " ", keyCode: 32 });

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");
  });

  it("should open and close mixing mouse click, Enter, and Space", async () => {
    const { getByRole } = render(<Main />);

    const baseDropdownHeader = getByRole("combobox");

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    await userEvent.click(baseDropdownHeader);

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "true");

    // Close
    fireEvent.keyDown(baseDropdownHeader, { key: "Enter", keyCode: 13 });

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");

    // Open
    fireEvent.keyDown(baseDropdownHeader, { key: " ", keyCode: 32 });

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "true");

    // Close
    await userEvent.click(baseDropdownHeader);

    expect(baseDropdownHeader).toHaveAttribute("aria-expanded", "false");
  });
});
