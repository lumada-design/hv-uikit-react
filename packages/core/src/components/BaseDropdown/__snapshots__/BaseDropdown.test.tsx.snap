// Vitest Snapshot v1

exports[`BaseDropDown > should render correctly 1`] = `
<div>
  <div
    style="width: 121px;"
  >
    <div
      class="css-1nukton-StyledRoot e1xgfj1t9"
    >
      <div
        aria-expanded="false"
        aria-label="Main sample"
        class=" css-c41kvr-StyledAnchor e1xgfj1t8"
        role="combobox"
        tabindex="-1"
      >
        <div
          aria-label="Main sample"
          class=" css-z15in3-StyledHeaderRoot e1xgfj1t7"
          role="textbox"
          tabindex="0"
        >
          <div
            class="css-1c8fpra-StyledSelection e1xgfj1t6"
          >
            <p
              class=" e1xgfj1t5 css-fi1crd-StyledPlaceholder"
              variant="body"
            >
              Placeholder...
            </p>
          </div>
          <div
            class="css-tgcv4t-StyledDropDownXS e1xgfj1t3 css-du18qm"
            name="DropDownXS"
          >
            <svg
              focusable="false"
              height="12"
              viewBox="0 0 12 12"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                fill="var(--colors-acce1)"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
